File: Python/sentimental-readability/readability.py
# Coleman Liau index == 0.0588 * L - 0.296 * S - 15.8
# where:
#   L == avg#of letters per 100 words
#   S == avg#of sentences per 100 words
""" Counts the number of letter, words, and sentences, and prints as output "Grade X" where `X` is
    the grade level computed by the Coleman-Liau formula, rounded to the nearest integer.
    If the resulting index number is 16 or higher the program outputs `"Grade 16+"`, and likewise,
    if the resulting index number is less than 1, the program outputs `"Before Grade 1"`."""

from cs50 import get_string


def main():
    # Prompt user for text
    text = get_string("Text: ")
    # Call the index function
    index = calculate_index(text)
    # print Grade level
    print(f"{grade_level(index)}")


def calculate_index(text):
    # Count the #of words
    words = len(text.split())
    # Count the #of letters (L)
    L = len([char for char in text if char.isalpha()]) * 100 / words
    # Count the #of sentences (S)
    S = len([char for char in text if char in [".", "!", "?"]]) * 100 / words
    # Calculate index:  0.0588 * L - 0.296 * S - 15.8
    index = round(0.0588 * L - 0.296 * S - 15.8)
    return index


def grade_level(index):
    if index > 16:
        level = "Grade 16+"
    elif index < 1:
        level = "Before Grade 1"
    else:
        level = f"Grade {index}"
    return level


main()

File: Arrays/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    string text = get_string("Text: ");

    printf("\n%i letters\n", count_letters(text));
    printf("%i words\n", count_words(text));
    printf("%i sentences\n", count_sentences(text));

    // Calculate Coleman-Liau index
    float L = (count_letters(text) * 100.0 / count_words(text));   // L = avg# of Letters / 100 words
    float S = (count_sentences(text) * 100.0 / count_words(text)); // S = avg# of Sentences / 100 words
    int index = round((0.0588 * L) - (0.296 * S) - 15.8);

    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }

    return 0;
}

int count_letters(string text)
{
    int count_letters = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (isalpha(text[i]))
        {
            count_letters++;
        }
    }
    return count_letters;
}

int count_words(string text)
{
    int count_words = 1;
    for (int i = 0; i < strlen(text); i++)
    {
        if (isspace(text[i]))
        {
            count_words++;
        }
    }
    return count_words;
}

int count_sentences(string text)
{
    int count_sentences = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count_sentences++;
        }
    }
    return count_sentences;
}
